File: Solution Kruskal's Algorithm.js

/**
 * @param {number[][]} points
 * @return {number}
 */
var minCostConnectPoints = function (points) {
    if (points.length < 2) {
        return 0;
    }
    const totalPoints = points.length;
    const totalEdges = (totalPoints * (totalPoints - 1)) / 2;

    const edges = new Array(totalEdges).fill(null);
    initializeArrayEdges(points, edges, totalPoints);
    edges.sort((x, y) => x.distance - y.distance);

    return calculateMinimumCostConnectAllPoints(edges, totalEdges, totalPoints);
};

/**
 * @param {Edge[]} edges
 * @param {number} totalEdges 
 * @param {number} totalPoints 
 * @return {number}
 */
function calculateMinimumCostConnectAllPoints(edges, totalEdges, totalPoints) {

    const unionFind = new UnionFind(totalPoints);
    let countEdgesInMinSpanningTree = 0;
    let minCostConnectAllPoints = 0;

    for (let i = 0; i < totalEdges && countEdgesInMinSpanningTree < totalPoints - 1; ++i) {
        let parentFirst = unionFind.findParent(edges[i].firstPoint);
        let parentSecond = unionFind.findParent(edges[i].secondPoint);

        if (parentFirst !== parentSecond) {
            unionFind.joinByRank(parentFirst, parentSecond);
            ++countEdgesInMinSpanningTree;
            minCostConnectAllPoints += edges[i].distance;
        }
    }
    return minCostConnectAllPoints;
}

/**
 * @param {number[][]} points 
 * @param {Edge[]} edges
 * @param {number} totalPoints 
 * @return {void}
 */
function initializeArrayEdges(points, edges, totalPoints) {
    let index = 0;
    for (let i = 0; i < totalPoints; ++i) {
        for (let j = i + 1; j < totalPoints; ++j) {
            edges[index++] = new Edge(i, j, manhattanDistance(points[i], points[j]));
        }
    }
}

/**
 * @param {number} firstPoint 
 * @param {number} secondPoint 
 * @return {number}
 */
function manhattanDistance(firstPoint, secondPoint) {
    return Math.abs(firstPoint[0] - secondPoint[0]) + Math.abs(firstPoint[1] - secondPoint[1]);
}

/**
 * @param {number} firstPoint 
 * @param {number} secondPoint 
 * @param {number} distance 
 */
function Edge(firstPoint, secondPoint, distance) {
    this.firstPoint = firstPoint;
    this.secondPoint = secondPoint;
    this.distance = distance;
}

class UnionFind {

    constructor(totalPoints) {
        this.parent = Array.from(Array(totalPoints).keys());
        this.rank = new Array(totalPoints).fill(0);
    }

    findParent(index) {
        if (this.parent[index] !== index) {
            this.parent[index] = this.findParent(this.parent[index]);
        }
        return this.parent[index];
    }

    joinByRank(first, second) {
        if (this.rank[first] > this.rank[second]) {
            this.parent[second] = first;
        } else if (this.rank[first] < this.rank[second]) {
            this.parent[first] = second;
        } else {
            this.parent[second] = first;
            this.rank[first]++;
        }
    }
}

File: Solution Kruskal's Algorithm.java

import java.util.Arrays;
import java.util.stream.IntStream;

public class Solution {

    private static record Edge(int firstPoint, int secondPoint, int distance) {}

    public int minCostConnectPoints(int[][] points) {
        if (points.length < 2) {
            return 0;
        }
        final int totalPoints = points.length;
        final int totalEdges = (totalPoints * (totalPoints - 1)) / 2;

        Edge[] edges = new Edge[totalEdges];
        initializeArrayEdges(points, edges, totalPoints);
        Arrays.sort(edges, (x, y) -> x.distance - y.distance);

        return calculateMinimumCostConnectAllPoints(edges, totalEdges, totalPoints);
    }

    private int calculateMinimumCostConnectAllPoints(Edge[] edges, int totalEdges, int totalPoints) {

        UnionFind unionFind = new UnionFind(totalPoints);
        int countEdgesInMinSpanningTree = 0;
        int minCostConnectAllPoints = 0;

        for (int i = 0; i < totalEdges && countEdgesInMinSpanningTree < totalPoints - 1; ++i) {
            int parentFirst = unionFind.findParent(edges[i].firstPoint);
            int parentSecond = unionFind.findParent(edges[i].secondPoint);

            if (parentFirst != parentSecond) {
                unionFind.joinByRank(parentFirst, parentSecond);
                ++countEdgesInMinSpanningTree;
                minCostConnectAllPoints += edges[i].distance;
            }
        }
        return minCostConnectAllPoints;
    }

    private void initializeArrayEdges(int[][] points, Edge[] edges, int totalPoints) {
        int index = 0;
        for (int i = 0; i < totalPoints; ++i) {
            for (int j = i + 1; j < totalPoints; ++j) {
                edges[index++] = new Edge(i, j, manhattanDistance(points[i], points[j]));
            }
        }
    }

    private int manhattanDistance(int[] firstPoint, int[] secondPoint) {
        return Math.abs(firstPoint[0] - secondPoint[0]) + Math.abs(firstPoint[1] - secondPoint[1]);
    }
}

class UnionFind {

    private final int[] parent;
    private final int[] rank;

    UnionFind(int totalPoints) {
        parent = IntStream.range(0, totalPoints).toArray();
        rank = new int[totalPoints];
    }

    protected int findParent(int index) {
        if (parent[index] != index) {
            parent[index] = findParent(parent[index]);
        }
        return parent[index];
    }

    protected void joinByRank(int first, int second) {
        if (rank[first] > rank[second]) {
            parent[second] = first;
        } else if (rank[first] < rank[second]) {
            parent[first] = second;
        } else {
            parent[second] = first;
            rank[first]++;
        }
    }
}
